docs/design.md

```diff
@@ -0,0 +1,431 @@
+
+<!-- TODO: Add example(s) maybe simple + an example with source layering? -->
+
+### Control Documents
+
+Control documents (documents which have `metadata.schema=metadata/Control/v1`),
+are special, and are used to control the behavior of Deckhand at runtime.  Only
+the following types control documents are allowed.
+
+#### DataSchema
+
+`DataSchema` documents are used by various services to register new schemas
+that Deckhand can use for validation. No `DataSchema` documents with names
+beginning with `deckhand/` or `metadata/` are allowed.
+
+<!-- TODO: give valid, tiny schema as example -->
+
+```yaml
+---
+schema: deckhand/DataSchema/v1
+metadata:
+  schema: metadata/Control/v1
+  name: promenade/Node/v1  # Specifies the documents to be used for validation.
+  labels:
+    application: promenade
+data:  # Valid JSON Schema is expected here.
+  $schema: http://blah
+```
+
+#### LayeringPolicy
+
+Only one `LayeringPolicy` document can exist within the system at any time.
+It is an error to attempt to insert a new `LayeringPolicy` document if it has
+a different `metadata.name` than the existing document. If the names match,
+it is treated as an update to the existing document.
+
+This document defines the strict order in which documents are merged together
+from their component parts. It should result in a validation error if a
+document refers to a layer not specified in the `LayeringPolicy`.
+
+```yaml
+---
+schema: deckhand/LayeringPolicy/v1
+metadata:
+  schema: metadata/Control/v1
+  name: layering-policy
+data:
+  layerOrder:
+    - global
+    - site-type
+    - region
+    - site
+    - force
+```
+
+#### ValidationPolicy
+
+Unlike `LayeringPolicy`, many `ValidationPolicy` documents are allowed. This
+allows services to check whether a particular revision (described below) of
+documents meets a configurable set of validations without having to know up
+front the complete list of validations.
+
+Deckhand provides validation of all concrete documents given their schema
+under the name `deckhand-schema-validation`. Since validations may indicate
+interactions with external and changing circumstances, an optional
+`expiresAfter` key may be specified for each validation as an ISO8601
+duration. If no `expiresAfter` is specified, a successful validation does not
+expire.
+
+```yaml
+---
+schema: deckhand/ValidationPolicy/v1
+metadata:
+  schema: metadata/Control/v1
+  name: site-deploy-ready
+data:
+  validations:
+    - name: deckhand-schema-validation
+    - name: drydock-site-validation
+      expiresAfter: P1W
+    - name: promenade-site-validation
+      expiresAfter: P1W
+    - name: armada-deployability-validation
+```
+
+### Provided Utility Document Kinds
+
+These are documents that use the `Document` metadata schema, but live in the
+`deckhand` namespace.
+
+#### Certificate
+
+```yaml
+---
+schema: deckhand/Certificate/v1
+metadata:
+  schema: metadata/Document/v1
+  name: application-api
+  storagePolicy: cleartext
```

Comment:
Well, it's not a secret, so it wouldn't have to be, but if policy required it, it would be fine to do here.  This is configurable per document.